drivers/power/adp1050/adp1050.c

```diff
@@ -0,0 +1,1250 @@
+*/
+int adp1050_set_vin(struct adp1050_desc *desc, int16_t mantissa, int8_t exp,
+		    bool state_on)
+{
+	uint16_t val;
+
+	if (mantissa > ADP1050_MANT_MAX || exp > ADP1050_EXP_MAX)
+		return -EINVAL;
+
+	val = no_os_field_prep(ADP1050_EXP_MASK, exp) | mantissa;
+
+	return adp1050_write(desc, state_on ? ADP1050_VIN_ON : ADP1050_VIN_OFF,
+			     val, 2);
+}
+
+/**
+ * @brief Set current limit fast fault for overcurrent protection and current
+ * 	  limit cycle-by-cycle reference
+ * @param desc - ADP1050_device_descriptor.
+ * @param fast_fault - Fast fault limit value for N counter.
+ * @param climit_ref - Current limit cylcle-by-cycle reference.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int adp1050_set_climit(struct adp1050_desc *desc,
+		       enum adp1050_iin_oc_fast_fault fast_fault,
+		       enum adp1050_climit_ref climit_ref)
+{
+	int ret;
+
+	ret = adp1050_write(desc, ADP1050_IIN_OC_FAST_FAULT_LIMIT,
+			    (uint16_t)fast_fault, 1);
+	if (ret)
+		return ret;
+
+	return adp1050_write(desc, ADP1050_CS1_CBC_CURR_LIM_REF,
+			     (uint16_t)climit_ref, 1);
+}
+
+/**
+ * @brief Set current sense settings
+ * @param desc - ADP1050 device descriptor
+ * @param leading_edge - Leading edge blanking time.
+ * @param debounce - Current sense cycle-by-cycle current-limit debounce time.
+ * @param ignore - CS1 overcurrent protection comparator to be ignored.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int adp1050_set_cs1_settings(struct adp1050_desc *desc,
+			     enum adp1050_cs1_leading_edge leading_edge,
+			     enum adp1050_cs1_debounce debounce,
+			     bool ignore)
+{
+	uint16_t val;
+
+	if (ignore)
+		return adp1050_write(desc, ADP1050_CS1_CBC_CURR_LIM_SETTINGS,
+				     ADP1050_CS1_CLIMIT_COMP_IGNORE, 1);
+
+	val = no_os_field_prep(ADP1050_CS1_LEADING_EDGE_MASK, leading_edge) |
+	      no_os_field_prep(ADP1050_CS1_DEBOUNCE_MASK, debounce);
+
+	return adp1050_write(desc, ADP1050_CS1_CBC_CURR_LIM_SETTINGS, val, 1);
+}
+
+/**
+ * @brief Set requested ADP1050 channel in Open Loop operation mode.
+ * @param desc - ADP1050 device descriptor.
+ * @param rising_edge - Rising edge timing.
+ * @param falling_edge - Falling edge timing.
+ * @param chan - Channel t be set in open-loop.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int adp1050_set_open_loop(struct adp1050_desc *desc, uint8_t rising_edge,
+			  uint8_t falling_edge, enum adp1050_channel chan)
+{
+	int ret;
+
+	if ((chan == ADP1050_OUTA || chan == ADP1050_OUTB)
+	    && falling_edge < rising_edge)
+		return -EINVAL;
+
+	if ((chan == ADP1050_SR1 || chan == ADP1050_SR2) && rising_edge < falling_edge)
+		return -EINVAL;
+
+	ret = adp1050_set_pwm(desc, ADP1050_DISABLE_ALL, desc->freq);
+	if (ret)
+		return ret;
+
+	ret = adp1050_write(desc, ADP1050_FEEDFORWARD_SS_FILTER_GAIN,
+			    ADP1050_CL_FEEDFORWARD_DISABLE, 1);
+	if (ret)
+		return ret;
+
+	switch (chan) {
```

Comment:
this switch contain on each case a lot of duplicated code, I think it could be rewritten in 2 functions and call them for first 2 cases, and next 2 cases respectively